### DISCUSSION TITLE
Allow for more than 3x2pt in the systematics factory interface

### DISCUSSION THREAD

**Discussion by arthurmloureiro** [LSSTDESC/firecrown#464] [Ideas]
Hi Firecrown team,

This is something I have been discussing with @vitenti : The current systematics factories work amazingly well to generate likelihoods from metadata and yamls but at the moment it seems like it is quite tailored to only build 3x2pt likelihoods. It does not seem like an extra lens sample could be added (from a spectroscopic sample, for example) nor that something like CMB cross-correlations could be added to the `TwoPoint` factories.

For example:
```python
all_two_point_functions = tp.TwoPoint.from_metadata(
    metadata_seq=all_two_point_cells,
    wl_factory=wlf,
    nc_factory=ncf,
)
```
Allows only for a single `nc_factory` to be passed. 

It would be interesting to allow for more and different  factories to be parsed to this function so multi-tracer analysis are also supported.

@marcpaterno @vitenti let me know if this should be an issue instead :)

**Comment by vitenti** [comment]
Hello @arthurmloureiro ,

For the case "an extra lens sample could be added", we discussed a solution where the user has better control on the factory behavior based on the bin characteristics. For example, the PhotozShift would not be applied if one particular bin in your analysis is spectroscopic. We think that could solve the problem of combining photometric and spectroscopic data. What do you think?

About the not yet implemented observables, do you have any particular concrete example in mind? For cross correlation, we have pyccl.CMBLensingTracer and pyccl.tSZTracer available in CCL that are not yet supported by Firecrown.

**Reply by arthurmloureiro** [reply]
Hi @vitenti !

Yes, I think that could work!
So the idea is that we would have a PhotozShift factory with the same number of bins as the Photo-z tracer and then a tracer without these factories in a way that we would still have:
```python
all_two_point_functions = tp.TwoPoint.from_metadata(
    metadata_seq=all_two_point_cells,
    wl_factory=wlf,
    nc_factory=ncf,
)
```
where both cases are included in the `nc_factory`?

For the CMB Tracers, the only other one that we have in CCL but not yet in Firecrown that I can think of is ISW.

Would these be implemented inside this `tp.TwoPoint.from_metadata` too? (Happy to help btw!)

**Comment by tilmantroester** [comment]
I think more flexibility for matching factories to tracers would be generally useful. I could also imagine different source galaxy samples with different intrinsic alignment systematics, for example. So rather than having one factory per tracer type, allow for one factory per group of tracers, where the group defaults to all tracers of the same type (e.g. weak lensing or number counts) but allows for other groupings as well (e.g. two groups of number count tracers with different photo-z systematics).

**Reply by vitenti** [reply]
@tilmantroester @arthurmloureiro 

We are working on a suitable design to address this. One option is to pass all the tracer information to the factory, allowing it to decide how to handle each case based on the provided data. In this approach, the factory would manage the complexity of selecting the appropriate systematic effect or model -- such as intrinsic alignment -- or, when necessary, return an `IdentitySystematic` (for cases where a particular model doesn't apply).

Another option is to implement a filter system that maps tracer characteristics to corresponding systematics. For example, a rule like `Galaxies.COUNTS + RedshiftMethod.PHOTOZ => PhotoZShiftFactory` would ensure that the `PhotoZShiftFactory` is only invoked for tracers with these specific attributes in their metadata.

What do you think about these approaches?

**Reply by tilmantroester** [reply]
To me the second approach sounds more flexible. Especially if there's a utility that makes specifying the filters easy. Something like tracers with "spec" in the sacc name use `spec_number_counts_factory`, while those with "lens" in the name use `photoz_number_counts_factory`. I think that would cover @arthurmloureiro use case. I think leaving building the filters to the user is fine.

**Reply by arthurmloureiro** [reply]
I am not fully sure how the filtering would work but feels to me this could also solve the issue where, for example, source and lenses share the same redshift systematics.

**Comment by arthurmloureiro** [comment]
Tagging @VGozzi here as he has a way of implementing  6x2pt with the current firecrown :)

**Comment by arthurmloureiro** [comment]
Any news on this? I can try to do a PR ...

I am still having all sorts of unreasonable errors from several checks within firecrown which I am not sure we should be checking for. Not sure why I cannot combine `galaxies.COUNTS` with `galaxies.SHEAR_PLUS` for example... Removing these restrictions and checks would allow us to use more than one tracer for the same observable, which would be incredibly useful.

**Comment by marcpaterno** [comment]
The items in this discussion are resolved in PR #512. Any addition discussion should be put into the comments on that PR.
